repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomFeedServiceImpl.scala

```diff
@@ -0,0 +1,230 @@
+
+  @PostConstruct
+  def init(): Unit = {
+    logger.info("Initializing and registering configuration listeners")
+    val xsdURL = getClass.getResource("/META-INF/schema/config/atom-feed-service.xsd")
+
+    configurationService.subscribeTo(
+      DEFAULT_CONFIG,
+      xsdURL,
+      AtomFeedServiceConfigurationListener,
+      classOf[AtomFeedServiceConfigType]
+    )
+    configurationService.subscribeTo(
+      SYSTEM_MODEL_CONFIG,
+      SystemModelConfigurationListener,
+      classOf[SystemModel]
+    )
+  }
+
+  @PreDestroy
+  def destroy(): Unit = {
+    logger.info("Unregistering configuration listeners and shutting down service")
+    configurationService.unsubscribeFrom(DEFAULT_CONFIG, AtomFeedServiceConfigurationListener)
+    configurationService.unsubscribeFrom(SYSTEM_MODEL_CONFIG, SystemModelConfigurationListener)
+
+    actorSystem.shutdown()
+  }
+
+  override def registerListener(feedId: String, listener: AtomFeedListener): String = {
+    val listenerId = java.util.UUID.randomUUID().toString
+    logger.info("Attempting to register a listener with id: " + listenerId + " to feed with id: " + feedId)
+
+    val notifierManager = feedActors.get(feedId) match {
+      case Some(FeedActorPair(manager, _)) =>
+        manager
+      case None =>
+        val manager = actorSystem.actorOf(Props[NotifierManager], name = feedId + NOTIFIER_MANAGER_TAG)
+        feedActors = feedActors + (feedId -> FeedActorPair(manager, None))
+
+        if (isServiceEnabled) {
+          manager ! ServiceEnabled
+        }
+
+        manager
+    }
+
+    notifierManager ! AddNotifier(listenerId, listener)
+
+    listenerNotifierManagers = listenerNotifierManagers + (listenerId -> notifierManager)
+
+    listenerId
+  }
+
+  override def unregisterListener(listenerId: String): Unit = {
+    logger.info("Attempting to unregister the listener with id: " + listenerId)
+
+    listenerNotifierManagers.get(listenerId) match {
+      case Some(notifierManager) =>
+        notifierManager ! RemoveNotifier(listenerId)
+        listenerNotifierManagers = listenerNotifierManagers - listenerId
+      case None =>
+        logger.debug("Listener not registered: " + listenerId)
+    }
+  }
+
+  case class FeedActorPair(notifierManager: ActorRef, feedReader: Option[ActorRef])
+
+  private object AtomFeedServiceConfigurationListener extends UpdateListener[AtomFeedServiceConfigType] {
+    private var initialized = false
+
+    override def configurationUpdated(configurationObject: AtomFeedServiceConfigType): Unit = {
+      synchronized {
+        logger.trace("Service configuration updated")
+        initialized = true
+
+        configurationObject.getFeed foreach { feedConfig =>
+          val notifierManager = feedActors.get(feedConfig.getId) match {
+            case Some(actorPair) =>
+              // todo: only replace the current feed reader if the configuration has changed
```

Comment:
Oh, fair point. I want to say I thought about that, because it seems pretty obvious... let me look at the code again.

Ah, okay. So I think this was my thinking... If we send the PoisonPill message, then the current FeedReader will finish any ReadFeed messages that came in prior to the PoisonPill. If that's the case, then the old reader should read all of the entries in the Feed, and when the new FeedReader starts, it will mark the most recent entry as the high water mark and start reading from there.

I guess there should be some concern over the sychronization... that's hard since we don't know exactly when the old FeedReader will stop. The old FeedReader still reading could lead to duplicate events, or even missed events if the old FeedReader dies immediately and a new entry is put on the feed before the new FeedReader reads for the first time.

If we hard stop the old FeedReader rather than using a PoisonPill, we know exactly when it stops, but not on which entry it stops.

So I guess the solution is to only replace a FeedReader if its configuration has changed. That solves the problem with replacing the FeedReader. How would we check the configuration delta? We have to store the old configuration in some form. We could store the configuration for a FeedReader within the FeedReader itself, but then we have to communicate with the ActorSystem from outside of it, which isn't very clean. Actually, I think the easiest way would be to store the entire old configuration object (or, at least, the portion pertaining to feeds). Then just check every field against itself in the new feed configuration. Does that sound reasonable?